toolchain/sem_ir/value_stores.h

```diff
@@ -146,17 +146,35 @@ class ConstantValueStore : public Yaml::Printable<ConstantValueStore> {
   llvm::SmallVector<ConstantId, 0> values_;
 };
 
-// Provides storage for instructions representing global constants.
+// Provides storage for instructions representing deduplicated global constants.
 class ConstantStore {
  public:
-  // Add a constant instruction.
-  auto Add(InstId inst_id) -> void { values_.push_back(inst_id); }
+  explicit ConstantStore(SemIR::File* sem_ir, llvm::BumpPtrAllocator& allocator)
+      : allocator_(&allocator), constants_(sem_ir) {}
 
-  auto array_ref() const -> llvm::ArrayRef<InstId> { return values_; }
-  auto size() const -> int { return values_.size(); }
+  // Adds a new constant instruction, or gets the existing constant with this
+  // value. Returns the instruction ID and a bool indicating whether a new
+  // instruction was added.
+  auto GetOrAdd(Inst inst) -> std::pair<SemIR::InstId, bool>;
+
+  auto vector() const -> std::vector<InstId>;
+  auto size() const -> int { return constants_.size(); }
 
  private:
-  llvm::SmallVector<InstId> values_;
+  // TODO: We store two copies of each constant instruction: one in insts() and
+  // one here. We could avoid one of those copies and store just an InstId here,
+  // at the cost of some more indirection when recomputing profiles during
+  // lookup. Once we have a representative data set, we should measure the
+  // impact on compile time from that change.
+  struct ConstantNode : llvm::FoldingSetNode {
+    Inst inst;
+    InstId inst_id;
+
+    auto Profile(llvm::FoldingSetNodeID& id, SemIR::File* sem_ir) -> void;
```

Comment:
Style: functions before members.
```suggestion
    auto Profile(llvm::FoldingSetNodeID& id, SemIR::File* sem_ir) -> void;

        Inst inst;
    InstId inst_id;
```

Reply:
I'd prefer to keep this as-is. Style [allows either order](https://google.github.io/styleguide/cppguide.html#Declaration_Order) for structs, and data-first is more consistent with classes -- at least for aggregate structs (those without constructors) -- because it puts construction information at the start.

Reply:
It would really be nice to be internally consistent within Carbon...

Reply:
I did some checking, and we're currently not especially consistent. For aggregates with both fields and methods, we have an approximately 2:1 ratio of methods first versus fields first (though it's closer to being 1:1 if you count a cluster of types using the same convention as one):

- Fields first: `Check::ConversionTarget`, `StringLiteral::Introducer`, `CharSet` (string_literal.cpp), `RandomSourceOptions` (tokenized_buffer_benchmark.cpp), `Step` (in `File::StringifyTypeExpr`), `SemIR::InitRepr`, `InstNamer::Namespace`, `ConstantStore::ConstantNode`

- Methods first: `Check::DeclNameStack::NameContext`, `Driver::{CompileOptions,Options}`, `NibbleLUT` (lex.cpp), `ExpectedTokens` (tokenized_buffer_test_helpers.h), `TokenizedBuffer::PrintWidths`, `Parse::Context::StateStackEntry`, `SemIR::{BindNameInfo,Function,Class,Interface}`, `SemIR::ValueRepr`, `SemIR::TypeInfo`, `SemIR::NameScope`, `Printer` (yaml_test_helpers.cpp)

(Granted, I added most of the types in the first category and you added most of the types in the second!) Personally I think the more important consistency is putting the most important information about a type first, rather than putting either fields or methods first, so I would prefer that we allow the author of the code to choose the option that best communicates their intent rather than enforcing one option or the other.